File: springboot-backend/src/main/java/com/content/publishing/controller/AdminController.java
package com.content.publishing.controller;

import com.content.publishing.dto.AuthDtos.RoleUpdateRequest;
import com.content.publishing.dto.AuthDtos.UserResponse;
import com.content.publishing.service.UserService;
import java.util.List;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * Admin-only endpoints.
 * The FastAPI gateway enforces ADMIN role before forwarding requests here.
 */
@RestController
@RequestMapping("/api/admin")
public class AdminController {

    private final UserService userService;

    public AdminController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping("/users")
    public List<UserResponse> listUsers() {
        return userService.listAllUsers();
    }

    @PutMapping("/users/{userId}/role")
    public UserResponse updateRole(
            @PathVariable Long userId,
            @RequestBody RoleUpdateRequest request
    ) {
        return userService.updateRole(userId, request.role());
    }
}

File: fastapi-gateway/controllers/admin_controller.py
from fastapi import APIRouter, Depends

from controllers import RoleUpdateRequest, current_user, require_admin, spring_request

router = APIRouter(prefix="/admin", tags=["admin"])


@router.get("/users")
async def list_users(user: dict = Depends(current_user)):
    """List all users — ADMIN only."""
    require_admin(user)
    return await spring_request("GET", "/api/admin/users")


@router.put("/users/{user_id}/role")
async def update_user_role(
    user_id: int,
    request: RoleUpdateRequest,
    user: dict = Depends(current_user)
):
    """Change a user's role — ADMIN only."""
    require_admin(user)
    return await spring_request(
        "PUT", f"/api/admin/users/{user_id}/role",
        json=request.model_dump()
    )
